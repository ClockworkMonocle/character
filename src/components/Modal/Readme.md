For when you want to display something on top of something else. This component only allows for positioning of the modal, you're on your own for the styles!

Users can press the `ESC` key to invoke `onDismiss` as well.

**NOTE** Modals must start with `active` set to `false`, otherwise the modal will not render properly.

Default usage:

```js
require('../Button');
require('../Type');

const modalStyle = {
  backgroundColor: 'var(--white)',
  maxHeight: '400px',
  padding: '8px',
  textAlign: 'center',
};
const toggleOpen = () => setState({ open: !state.open });

<div>
  <Button key="mtb" onClick={toggleOpen} variant="pill" color="proficient">Toggle</Button>
  <Modal key="mmm" onDismiss={toggleOpen} active={state.open}>
    <div style={modalStyle}>
      <Type variant="display-2" color="attack-color">Engage!</Type>
    </div>
  </Modal>
</div>
```

Can also fiddle with the placement for easily made Popup and Drawer components.

```js
require('../Button');
require('../Type');

const drawerStyle = {
  height: '100%',
  maxWidth: '180px',
  backgroundColor: 'var(--dark)',
  padding: '8px',
};

const bottomStyle = {
  maxWidth: '220px',
  backgroundColor: 'var(--dark)',
  padding: '8px',
  margin: '0 auto',
}

const toggleBottom = () => setState({ openBottom: !state.openBottom });
const toggleTop = () => setState({ openTop: !state.openTop });
const toggleLeft = () => setState({ openLeft: !state.openLeft });
const toggleRight = () => setState({ openRight: !state.openRight });

<div>
  <Button key="mta" onClick={toggleBottom} variant="pill" color="proficient">Toggle bottom</Button>
  <Button key="mtb" onClick={toggleLeft} variant="pill" color="proficient">Toggle left</Button>
  <Button key="mtc" onClick={toggleRight} variant="pill" color="proficient">Toggle right</Button>
  <Button key="mtd" onClick={toggleTop} variant="pill" color="proficient">Toggle top</Button>
  <Modal key="mmt" onDismiss={toggleTop} active={state.openTop} from="top">
    <div style={bottomStyle}>
      <Type variant="heading-1" color="attack-color">From the top of the screen</Type>
    </div>
  </Modal>
  <Modal key="mmp" onDismiss={toggleBottom} active={state.openBottom} from="bottom">
    <div style={bottomStyle}>
      <Type variant="heading-1" color="attack-color">From the bottom of the screen</Type>
    </div>
  </Modal>
  <Modal key="mml" onDismiss={toggleLeft} active={state.openLeft} from="left">
    <div style={drawerStyle}>
      <Type variant="heading-1" color="attack-color">From the left of the screen</Type>
    </div>
  </Modal>
  <Modal key="mmr" onDismiss={toggleRight} active={state.openRight} from="right">
    <div style={drawerStyle}>
      <Type variant="heading-1" color="attack-color">From the right of the screen</Type>
    </div>
  </Modal>
</div>
```
