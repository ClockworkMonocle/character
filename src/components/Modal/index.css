.modal {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 1337;
}

.modal_overlay {
  position: absolute;
  top: 0;
  height: 100%;
  width: 100%;
}

.modal_overlay--events {
  background-color: rgba(0, 0, 0, 0.75);
  display: flex;
  align-items: center;
  justify-content: center;
  opacity: 0;
  transform: scale(1.1);
  transition: all 300ms ease;
}

.modal_overlay--enter {
  opacity: 1;
  transform: scale(1);
}
