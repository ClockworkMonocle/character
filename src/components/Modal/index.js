import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { createPortal } from 'react-dom';
import cn from 'classnames';

import './index.css';

export default class Modal extends Component {
  static propTypes = {
    /** Set to true to show the modal */
    active: PropTypes.bool,
    /** Set where the modal enters the window from */
    from: PropTypes.oneOf(['middle', 'top', 'bottom', 'left', 'right']),
    onDismiss: PropTypes.func.isRequired,
  };

  static defaultProps = {
    active: false,
    from : 'middle',
  };

  overlayRef = null;

  constructor(props) {
    super(props);

    this.el = document.createElement('div');
    this.el.classList.add('modal');

    this.state = {
      shouldRenderPortal: props.active,
      hasEntered: false,
    };
  }

  componentWillReceiveProps(nextProps) {
    // if we need to take action...
    if (this.props.active !== nextProps.active) {

      // if we should be opening...
      if (nextProps.active) {
        this.addtoDom();

        this.setState(
          { shouldRenderPortal: true },
          () => setTimeout(
            () => this.setState({ hasEntered: true }),
            100
          )
        );
      }

      // if we should be closing...
      else {
        this.setState(
          { hasEntered: false },
          () => setTimeout(
            () => this.setState(
              { shouldRenderPortal: false },
              this.removeFromDom
            ),
            300
          )
        );
      }
    }
  }

  componentWillUnmount() {
    this.removeFromDom();
    this.el = null;
  }

  handleClick = e => e.target === this.overlayRef && this.props.onDismiss();
  handleEsc = e => e.keyCode === 27 && this.props.onDismiss();

  addtoDom = () => {
    document.body.style.overflow = 'hidden';
    document.body.appendChild(this.el);
    window.addEventListener('keydown', this.handleEsc);
  }

  removeFromDom = () => {
    document.body.style.overflow = 'inherit';
    if (document.body.contains(this.el)) {
      document.body.removeChild(this.el);
    }
    window.removeEventListener('keydown', this.handleEsc);
  }

  render() {
    const { from } = this.props;
    const { hasEntered, shouldRenderPortal } = this.state;
    const css = cn(
      'modal_overlay',
      from === 'middle' && `modal_overlay--middle`,
      hasEntered && 'modal_overlay--enter'
    );

    const contentCss = cn(
      'modal_content',
      `modal_content--${from}`,
      hasEntered && `modal_content--enter`,
    );

    if (!shouldRenderPortal) return null;

    return createPortal(
      (<div ref={e => (this.overlayRef = e)} className={css} onClick={this.handleClick}>
        <div className={contentCss}>
          {this.props.children}
        </div>
      </div>),
      this.el
    );
  }
}
