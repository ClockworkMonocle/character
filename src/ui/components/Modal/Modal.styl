.modal-overlay
  position fixed
  opacity 0
  z-index 11
  top 0
  bottom 0
  right 0
  left 0
  background-color rgba(0,0,0, 0.65);
  overflow hidden
  line-height 14px
  transition opacity 300ms ease
  
  display flex
  align-items center
  justify-content center
  
  &.modal-overlay-active
    opacity 1
  
  .modal-content-container
    transform scale(1)
    max-width 500px
    flex 1
    max-height 95%
    margin 2.5% 15px
    overflow-y scroll
    -webkit-overflow-scrolling touch
    border-radius $base-border-radius
    transition transform 300ms ease
    
    &.modal-content-transitioning
      transform scale(1.1)
    
    &.modal-content-active
      transform scale(1)
      opacity 1
    
    .modal-header
      padding $base-padding
      background-color #333
      color $base-background-color
      border-bottom 2px solid #000
      border-top-left-radius $base-border-radius
      border-top-right-radius $base-border-radius
      
      h3
        margin 0
      
    .modal-content
      padding $base-padding
      background-color $base-background-color
      
      p
        line-height 1.4
      
    .modal-footer
      background-color $base-background-color
      border-top 1px solid $base-border-color
      text-align center
      
      button
        height 30px
        appearance none
        border 0
        display inline-block
        outline none
        
        &:disabled,
        &.disabled
          color $text-gray !important
          background-color $bg-gray
          
          &:active
            background-color $bg-gray !important
        
        &:first-child:nth-last-child(1)
          width 100%
          
        &:first-child:nth-last-child(2),
        &:first-child:nth-last-child(2) ~ button
          width 50%
          
        &:first-child:nth-last-child(3),
        &:first-child:nth-last-child(3) ~ button
          width 33.333333333333%
          
        &:first-child:nth-last-child(4),
        &:first-child:nth-last-child(4) ~ button
          width 25%
        
        &:active
          background-color #ddd
        
        p
          font-size 13px
        