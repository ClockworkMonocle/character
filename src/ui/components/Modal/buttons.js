'use strict';

import React from 'react';
import Icon from '../Icon';

export function createSaveBtn(clickHandler) {
  return (
    <button 
      onClick={clickHandler}
      className='text-green'
    ><Icon icon='fa fa-pencil'/> Save</button>
  );
}

export function createEditBtn(clickHandler) {
  return (
    <button
      onClick={clickHandler}
      className='text-green'
    ><Icon icon='fa fa-pencil'/> Edit</button>
  );
}

export function createRemoveBtn(clickHandler) {
  return (
    <button
      onClick={clickHandler}
      className='text-red'
    ><Icon icon='fa fa-remove'/> Remove</button>
  );
}

export function createCancelBtn(clickHandler) {
  return (
    <button
      onClick={clickHandler}
      className='text-red'
    ><Icon icon='fa fa-remove'/> Cancel</button>
  );
}

export function createRestBtn(clickHandler) {
  return (
    <button
      onClick={clickHandler}
      className='text-green'
    ><Icon icon='fa fa-bed'/> Rest</button>
  );
}