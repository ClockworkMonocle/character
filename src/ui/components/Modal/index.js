

import PropTypes from 'prop-types';

import React, { Component } from 'react';
import ReactDOM from 'react-dom';

import cn from 'classnames';
import Portal from '../Portal';

// TODO: should require some constants to find config identifiers for html
// also see Popup/index.js
const appContainer = '.character-body';
const paneContainer = '.swiper-slide-active>.base-pane-container';

export default class extends React.Component {
  static displayName = 'Modal';

  static propTypes = {
    active: PropTypes.bool.isRequired,
    id: PropTypes.string.isRequired,
    content: PropTypes.element.isRequired,
    onDismiss: PropTypes.func.isRequired,
    overflowAppContainer: PropTypes.string,
    overflowPaneContainer: PropTypes.string,
  };

  static defaultProps = {
    active: false
  };

  state = {
    active: false,
    open: false,
    moving: false
  };

  componentWillReceiveProps(nextProps) {
    let _appContainer = nextProps.overflowAppContainer || appContainer;
    let _paneContainer = nextProps.overflowPaneContainer || paneContainer;

    if (nextProps.active !== this.props.active && nextProps.active === true) {
      document.querySelector(_appContainer).style.overflow = 'hidden';
      document.querySelector(_paneContainer).style.overflow = 'hidden';
      this.setState({ active: true, moving: true }, () => {
        setTimeout(() => {
          this.setState({ open: true, moving: false });
        }, 100);
      })
    }
    else if (nextProps.active !== this.props.active && nextProps.active === false) {
      this.setState({ open: false, moving: true }, () => {
        setTimeout(() => {
          document.querySelector(_appContainer).style.overflow = 'auto';
          document.querySelector(_paneContainer).style.overflow = 'auto';
          this.setState({ active: false, moving: false });
        }, 300);
      })
    }
  }

  dismiss = (ev) => {
    if (ev.target === ReactDOM.findDOMNode(this.refs.overlay)) {
      ev.preventDefault();
      ev.stopPropagation();

      this.props.onDismiss();
    }
  };

  _dismiss = () => {
    this.props.onDismiss();
  };

  componentWillUnmount() {
    let app = document.querySelector(appContainer);
    let pane = document.querySelector(paneContainer);

    if (app) {
      app.style.overflow = 'auto';
    }

    if (pane) {
      pane.style.overflow = 'auto';
    }
  }

  render() {
    let css = cn({
      'modal-overlay': true,
      'modal-overlay-active': this.state.open
    })

    let container = cn({
      'modal-content-container': true,
      'modal-content-transitioning': this.state.moving,
      'modal-content-active': this.state.open
    })

    return (this.state.active ?
      <Portal id={this.props.id}>
        <div ref='overlay' className={css} onClick={this.dismiss}>
          <div ref='content' className={container}>
            {React.cloneElement(this.props.content, { parentDismiss: this._dismiss })}
            { this.props.children }
          </div>
        </div>
      </Portal>
      : null)
  }
}
