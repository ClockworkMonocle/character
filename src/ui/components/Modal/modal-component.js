'use strict';

import React from 'react';
import ReactDOM from 'react-dom';
import cn from 'classnames';

export default React.createClass({
  displayName: 'Modal',


  getInitialState() {
    return {
      open: false
    }
  },


  getDefaultProps() {
    return {
      onDismiss: () => {},
      container: ''
    }
  },


  dismiss(ev) {
    if (ev.target === ReactDOM.findDOMNode(this.refs.overlay)) {
      ev.preventDefault();
      ev.stopPropagation();

      if (this.props.checkDismiss) {
        let result = this.props.checkDismiss();
        if (result) {
          this._close();
        }
      }
      else {
        this._close();    
      }
    }
  },


  _close() {
    this.setState({ open: false }, () => {
      document.querySelector(this.props.container).style.overflow = 'auto';
      setTimeout(() => {
        this.props.onDismiss();
      }, 300);
    })
  },


  componentDidMount() {
    document.querySelector(this.props.container).style.overflow = 'hidden';
    setTimeout(() => {
      this.setState({ open: true });
    }, 100);
  },


  componentWillUnmount() {
    document.querySelector(this.props.container).style.overflow = 'auto';
  },


  componentWillReceiveProps(nextProps) {
    if (nextProps.active !== undefined && nextProps.active === false) {
      this._close();
    }
  },


  render() {
    let css = cn({
      'modal-overlay': true,
      'modal-overlay-active': this.state.open
    })

    let container = cn({
      'modal-content-container': true,
      'modal-content-active': this.state.open
    })

    return (
      <div ref='overlay' className={css} onClick={this.dismiss}>
        <div ref='content' className={container}>
          {this.props.children}
        </div>
      </div>
    )
  }
})