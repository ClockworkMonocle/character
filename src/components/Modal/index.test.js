import React from 'react';
import { shallow, mount } from 'enzyme';
import Modal from './index';

jest.mock('react-dom', () => ({
  ...require.requireActual('react-dom'),
  createPortal: children => <div className="portal-mock">{children}</div>,
}));

test('<Modal /> renders default portal when active', () => {
  const testModal = (
    <Modal active onDismiss={() => {}}>
      <div>Heyo</div>
    </Modal>
  );
  expect(shallow(testModal)).toMatchSnapshot();
});

test('<Modal /> renders top portal when active', () => {
  const testModal = (
    <Modal active from="top" onDismiss={() => {}}>
      <div>from top</div>
    </Modal>
  );
  expect(shallow(testModal)).toMatchSnapshot();
});

test('<Modal /> renders bottom portal when active', () => {
  const testModal = (
    <Modal active from="bottom" onDismiss={() => {}}>
      <div>from bottom</div>
    </Modal>
  );
  expect(shallow(testModal)).toMatchSnapshot();
});

test('<Modal /> renders left portal when active', () => {
  const testModal = (
    <Modal active from="left" onDismiss={() => {}}>
      <div>from left</div>
    </Modal>
  );
  expect(shallow(testModal)).toMatchSnapshot();
});

test('<Modal /> renders right portal when active', () => {
  const testModal = (
    <Modal active from="right" onDismiss={() => {}}>
      <div>from right</div>
    </Modal>
  );
  expect(shallow(testModal)).toMatchSnapshot();
});

test('<Modal /> renders nothing when not active', () => {
  const testModal = (
    <Modal onDismiss={() => {}}>
      <div>not rendered</div>
    </Modal>
  );
  expect(shallow(testModal)).toMatchSnapshot();
});

test('<Modal /> renders active css', () => {
  window.setTimeout = fn => fn();
  const testModal = (
    <Modal onDismiss={() => {}}>
      <div>some cool modal body</div>
    </Modal>
  );

  const wrapper = mount(testModal);
  expect(wrapper.find('div.modal_overlay').exists()).toEqual(false);
  wrapper.setProps({ active: true });
  expect(wrapper.find('div.modal_overlay').exists()).toEqual(true);
  wrapper.update();
  expect(wrapper.find('div.modal_overlay--enter').exists()).toEqual(true);
});

test('<Modal /> opens and closes', () => {
  window.setTimeout = fn => fn();
  const testModal = (
    <Modal onDismiss={() => {}}>
      <div>some cool modal body</div>
    </Modal>
  );

  const wrapper = mount(testModal);
  expect(wrapper.find('div.modal_overlay').exists()).toEqual(false);
  wrapper.setProps({ active: true });
  expect(wrapper.find('div.modal_overlay').exists()).toEqual(true);
  wrapper.update();
  expect(wrapper.find('div.modal_overlay--enter').exists()).toEqual(true);
  wrapper.setProps({ active: false });
  wrapper.update();
  expect(wrapper.find('div.modal_overlay--enter').exists()).toEqual(false);
  expect(wrapper.find('div.modal_overlay').exists()).toEqual(false);
});

test('<Modal /> does nothing if given other props', () => {
  window.setTimeout = fn => fn();
  const testModal = (
    <Modal onDismiss={() => {}}>
      <div>some cool modal body</div>
    </Modal>
  );

  const wrapper = mount(testModal);
  expect(wrapper.find('div.modal_overlay').exists()).toEqual(false);
  wrapper.setProps({ active: true });
  expect(wrapper.find('div.modal_overlay').exists()).toEqual(true);
  wrapper.update();
  wrapper.setProps({ active: true });
  wrapper.update();
  expect(wrapper.find('div.modal_overlay').exists()).toEqual(true);
  expect(wrapper.find('div.modal_overlay--enter').exists()).toEqual(true);
});

test('<Modal /> invokes onDismiss when clicking on overlay', () => {
  const dismissMock = jest.fn();
  const testModal = (
    <Modal active onDismiss={dismissMock}>
      <div>some cool modal body</div>
    </Modal>
  );

  const wrapper = mount(testModal);
  expect(wrapper.find('div.modal_overlay').exists()).toEqual(true);
  wrapper.find('div.modal_overlay').simulate('click');
  expect(dismissMock.mock.calls.length).toEqual(1);
});

test('<Modal /> invokes onDismiss when pressing ESC key', () => {
  const dismissMock = jest.fn();
  const testModal = (
    <Modal onDismiss={dismissMock}>
      <div>some cool modal body</div>
    </Modal>
  );

  const wrapper = mount(testModal);
  wrapper.instance().handleEsc({ keyCode: 27 });
  expect(dismissMock.mock.calls.length).toEqual(1);
  wrapper.unmount();
});
