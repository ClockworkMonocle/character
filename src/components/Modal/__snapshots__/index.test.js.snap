// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> renders bottom portal when active 1`] = `
<div
  className="portal-mock"
>
  <div
    className="modal_overlay"
    onClick={[Function]}
  >
    <div
      className="modal_content modal_content--bottom"
    >
      <div>
        from bottom
      </div>
    </div>
  </div>
</div>
`;

exports[`<Modal /> renders default portal when active 1`] = `
<div
  className="portal-mock"
>
  <div
    className="modal_overlay modal_overlay--middle"
    onClick={[Function]}
  >
    <div
      className="modal_content modal_content--middle"
    >
      <div>
        Heyo
      </div>
    </div>
  </div>
</div>
`;

exports[`<Modal /> renders left portal when active 1`] = `
<div
  className="portal-mock"
>
  <div
    className="modal_overlay"
    onClick={[Function]}
  >
    <div
      className="modal_content modal_content--left"
    >
      <div>
        from left
      </div>
    </div>
  </div>
</div>
`;

exports[`<Modal /> renders nothing when not active 1`] = `""`;

exports[`<Modal /> renders right portal when active 1`] = `
<div
  className="portal-mock"
>
  <div
    className="modal_overlay"
    onClick={[Function]}
  >
    <div
      className="modal_content modal_content--right"
    >
      <div>
        from right
      </div>
    </div>
  </div>
</div>
`;

exports[`<Modal /> renders top portal when active 1`] = `
<div
  className="portal-mock"
>
  <div
    className="modal_overlay"
    onClick={[Function]}
  >
    <div
      className="modal_content modal_content--top"
    >
      <div>
        from top
      </div>
    </div>
  </div>
</div>
`;
